onadata/apps/api/viewsets/open_data_viewset.py

```diff
@@ -16,22 +17,92 @@
+from onadata.libs.serializers.data_serializer import TableauDataSerializer
 from onadata.libs.serializers.open_data_serializer import OpenDataSerializer
 from onadata.libs.utils.common_tools import json_stream
-from onadata.libs.utils.csv_builder import CSVDataFrameBuilder
+from onadata.libs.utils.common_tags import (ATTACHMENTS, NOTES, GEOLOCATION)
 
 BaseViewset = get_baseviewset_class()
 IGNORED_FIELD_TYPES = ['select one', 'select multiple']
 
+# index tags
+DEFAULT_OPEN_TAG = '['
+DEFAULT_CLOSE_TAG = ']'
+DEFAULT_INDEX_TAGS = (DEFAULT_OPEN_TAG, DEFAULT_CLOSE_TAG)
+
 
 def replace_special_characters_with_underscores(data):
     return [re.sub(r"\W", r"_", a) for a in data]
 
 
+def clean_data_fields(data, xform):
+    """
+    Streamlines the row header fields
+    with the column header fields for the same form.
+    """
+    result = []
+    if data:
+        headers = xform.get_headers()
+        tableau_headers = remove_metadata_fields(headers)
+        # Yield the rows
+        rows = (row for row in data)
+        for row in rows:
+            # Include fields from the column headers
+            diff = set(tableau_headers).difference(set(row))
+            left_out_fields = {
+                field: None for field in diff}
+            row.update(left_out_fields)
+            result.append(row)
+    yield result
+
+
+def remove_metadata_fields(data):
+    """
+    Clean out unneccessary metadata fields
+    """
+    for field in METADATA_FIELDS:
+        if field in data:
+            data.remove(field)
+    return list(data)
+
+
+def flatten_repeat_data(data):
+    """
+    Loop through instance data and flatten repeat data
+    given the key "children/details" and nested_key/
+    abbreviated xpath "children/details/immunization/polio_1",
+    generate ["children", index, "immunization/polio_1"]
+    """
+    result = []
+    index_tags = DEFAULT_INDEX_TAGS
+    for record in data:
+        flat_dict = {}
+        for (key, value) in viewitems(record):
+            if isinstance(value, list) and key not in [
+                    ATTACHMENTS, NOTES, GEOLOCATION]:
+                # Aiming this implementation specific to repeats
+                for index, item in enumerate(value):
+                    for (nested_key, nested_val) in viewitems(item):
+                        xpaths = [
+                            '{key}{open_tag}{index}{close_tag}'.format(
+                                key=nested_key.split('/')[0],
+                                open_tag=index_tags[0],
+                                index=index,
+                                close_tag=index_tags[1])] + [
+                                    nested_key.split('/')[1]]
+                        xpaths = "/".join(xpaths)
+                        flat_dict.update({xpaths: nested_val})
+            else:
+                flat_dict.update({key: value})
+        # Add each row to resultant data
+        result.append(flat_dict)
+    return result
```

Comment:
Suggestion: Lets combine clean_data_fields and flatten_repeat_data into one function. That way we only loop through row data once, cleaning and flattening the data where needed. We would also just have one generator function in this case which might be good ?